File: mixtures.py
import numpy as np
import torch
import sys
import time

def load_log_likelihoods_from_tsv(file_path):
    """
    Load log likelihoods and counts from a TSV file.

    Args:
        file_path (str): The path to the TSV file.

    Returns:
        log_likelihood_matrix (ndarray): The log likelihoods matrix (unique observations x groups).
        loglik_counts (ndarray): The counts of each unique observation.
    """
    data = np.loadtxt(file_path, delimiter='\t')
    loglik_counts = data[:, 0]
    log_likelihood_matrix = data[:, 1:]
    loglik_counts = np.log(loglik_counts)

    return log_likelihood_matrix, loglik_counts


def em_algorithm(log_likelihood_matrix, loglik_counts, alpha0, threshold=1e-6, max_iters=5000, use_gpu=True, data_type = torch.float32):
    """
    Perform the EM algorithm to infer the mixing proportions given log likelihoods and counts.

    Args:
        log_likelihood_matrix (ndarray or tensor): The log likelihoods matrix (unique observations x groups).
        loglik_counts (ndarray or tensor): The counts of each unique observation.
        threshold (float): The convergence threshold.
        max_iters (int): The maximum number of iterations.
        use_gpu (bool): Whether to use GPU.

    Returns:
        theta (ndarray): The inferred mixing proportions for each group.
    """

    device = torch.device("cuda" if torch.cuda.is_available() and use_gpu else "cpu")

    #print if using cuda
    if device == torch.device("cuda"):
        print("Using CUDA")

    log_likelihoods = torch.tensor(log_likelihood_matrix, dtype=data_type, device=device)
    loglik_counts = torch.tensor(loglik_counts, dtype=data_type, device=device)

    # print size of log_likelihoods by multiplying elements by data_type size
    print("Size of log_likelihoods: ", log_likelihoods.numel() * log_likelihoods.element_size())

    num_unique_obs, num_groups = log_likelihoods.shape

    # Pre-allocate tensor
    log_weighted_likelihoods = torch.empty(num_unique_obs, num_groups, dtype=data_type, device=device)

    # Initialize mixing proportions (theta values)
    #theta = torch.ones(num_groups, dtype=data_type, device=device) / num_groups  # uniform initialization
    theta = torch.tensor(alpha0, dtype=data_type, device=device)

    prev_loss = torch.tensor(float('inf'), dtype=data_type, device=device)

    threshold_tensor = torch.tensor(threshold, dtype=data_type, device=device)

    for iteration in range(max_iters):
        # E-step: Compute responsibilities
        log_weighted_likelihoods.copy_(log_likelihoods + torch.log(theta))
        log_sum_exp = torch.logsumexp(log_weighted_likelihoods, dim=1, keepdim=True)
        log_responsibilities = log_weighted_likelihoods.sub_(log_sum_exp)

        # M-step: Update theta values weighted by log counts
        log_weighted_responsibilities = log_responsibilities.add_(loglik_counts.unsqueeze(1))
        torch.exp_(log_weighted_responsibilities)
        new_theta = log_weighted_responsibilities.sum(dim=0) / torch.sum(torch.exp(loglik_counts))

        # Compute the log likelihood
        log_likelihood = torch.sum(torch.exp_(log_sum_exp.add_(loglik_counts.unsqueeze(1))))

        # Check for convergence
        loss = -log_likelihood
        if abs(prev_loss - loss) < threshold_tensor:
            break
        prev_loss.copy_(loss)

        # Update theta
        theta.copy_(new_theta)

    return theta.detach().cpu().numpy(), iteration + 1


if __name__ == '__main__':
    # Get filename from command line arguments
    if len(sys.argv) < 2:
        print("Usage: python mixtures.py <filename>")
        sys.exit(1)
    filename = sys.argv[1]

    # Time loading the log likelihoods
    start = time.time()

    # Load the log likelihoods and counts from the TSV file
    log_likelihood_matrix, loglik_counts = load_log_likelihoods_from_tsv(filename)

    end = time.time()

    print(f"Loaded log likelihoods in {end - start} seconds")

    # Uniform initialization of mixing proportions
    alpha0 = np.ones(log_likelihood_matrix.shape[1]) / log_likelihood_matrix.shape[1]
    
    # Time the EM algorithm (float32)
    start = time.time()

    # Run the EM algorithm to infer mixing proportions
    theta, num_iterations = em_algorithm(log_likelihood_matrix, loglik_counts, alpha0)

    end = time.time()

    print(f"Ran EM algorithm (float32) in {end - start} seconds")
    print(f"Converged in {num_iterations} iterations")
    
    # Time the EM algorithm (float64)
    start = time.time()

    # Run the EM algorithm to infer mixing proportions
    theta, num_iterations = em_algorithm(log_likelihood_matrix, loglik_counts, theta, threshold=1e-3, data_type=torch.float64)

    end = time.time()

    print(f"Ran EM algorithm (float64) in {end - start} seconds")
    print(f"Converged in {num_iterations} iterations")

    # print cuda max memory
    print("Max memory allocated: ", torch.cuda.max_memory_allocated())
    
    print("Mixing proportions (theta values):", theta)

File: mixtures.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <torch/torch.h>
#include <chrono>


// Function to load log likelihoods and counts from a TSV file
template <typename T>
void load_log_likelihoods_from_tsv(const std::string& file_path, std::vector<T>& log_likelihood_matrix, std::vector<T>& loglik_counts) {
    std::ifstream file(file_path);
    if (!file.is_open()) {
        throw std::runtime_error("Unable to open file");
    }

    std::string line;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::vector<T> row;
        std::string value;
        while (std::getline(iss, value, '\t')) {
            row.push_back(std::stof(value));
        }
        loglik_counts.push_back(std::log(row[0]));
        log_likelihood_matrix.insert(log_likelihood_matrix.end(), row.begin() + 1, row.end());
    }

    file.close();
}

// EM algorithm function
template <typename T>
std::pair<torch::Tensor, int> em_algorithm(const std::vector<T>& log_likelihood_matrix, const std::vector<T>& loglik_counts, torch::ScalarType dtype, T threshold = 1e-6, int max_iters = 5000, bool use_gpu = true) {

    torch::Device device = (torch::cuda::is_available() && use_gpu) ? torch::kCUDA : torch::kCPU;
    if (device == torch::kCUDA) {
        std::cout << "Using CUDA" << std::endl;
    }

    int num_rows = loglik_counts.size();
    int num_cols = log_likelihood_matrix.size() / num_rows;

    torch::Tensor log_likelihoods = torch::from_blob((T*)log_likelihood_matrix.data(), {num_rows, num_cols}, dtype).clone().to(device);
    torch::Tensor loglik_counts_tensor = torch::from_blob((T*)loglik_counts.data(), {num_rows}, dtype).clone().to(device);

    // pre allocate log_weighted_likelihoods
    torch::Tensor log_weighted_likelihoods = torch::empty({num_rows, num_cols}, dtype).to(device);

    torch::Tensor theta = torch::ones({num_cols}, dtype).to(device) / num_cols;
    torch::Tensor prev_loss = torch::tensor(std::numeric_limits<T>::infinity(), dtype).to(device);
    torch::Tensor threshold_tensor = torch::tensor(threshold, dtype).to(device);

    int iteration;
    for (iteration = 0; iteration < max_iters; ++iteration) {
        // E-step: Compute responsibilities
        log_weighted_likelihoods.copy_(log_likelihoods + torch::log(theta));
        torch::Tensor log_sum_exp = torch::logsumexp(log_weighted_likelihoods, 1, true);
        //torch::Tensor log_responsibilities = log_weighted_likelihoods.sub_(log_sum_exp);
        log_weighted_likelihoods.sub_(log_sum_exp);

        // M-step: Update theta values weighted by log counts
        //torch::Tensor log_weighted_responsibilities = log_responsibilities.add_(loglik_counts_tensor.unsqueeze(1));
        log_weighted_likelihoods.add_(loglik_counts_tensor.unsqueeze(1));
        //log_weighted_responsibilities.exp_();
        log_weighted_likelihoods.exp_();
        torch::Tensor new_theta = log_weighted_likelihoods.sum(0) / torch::exp(loglik_counts_tensor).sum();

        // Compute the log likelihood
        torch::Tensor log_likelihood = torch::sum((log_sum_exp.add_(loglik_counts_tensor.unsqueeze(1))).exp_());

        // Check for convergence
        torch::Tensor loss = -log_likelihood;
        if (torch::abs(prev_loss - loss).lt(threshold_tensor).item<bool>()) {
            break;
        }
        prev_loss = loss;

        // Update theta
        theta.copy_(new_theta);
    }

    return {theta.cpu(), iteration + 1};
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <filename> <float|double>" << std::endl;
        return 1;
    }
    std::string filename = argv[1];
    std::string precision = (argc > 2) ? argv[2] : "double";

    torch::ScalarType dtype;
    if (precision == "double") {
        dtype = torch::kFloat64;
    } else {
        dtype = torch::kFloat32;
    }

    if (precision == "double") {
        using T = double;

        std::vector<T> log_likelihood_matrix;
        std::vector<T> loglik_counts;

        auto start = std::chrono::high_resolution_clock::now();
        
        load_log_likelihoods_from_tsv(filename, log_likelihood_matrix, loglik_counts);

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> load_time = end - start;
        std::cout << "Loaded log likelihoods in " << load_time.count() << " seconds" << std::endl;

        torch::Tensor theta;
        int num_iterations;

        start = std::chrono::high_resolution_clock::now();

        auto result = em_algorithm(log_likelihood_matrix, loglik_counts, dtype, 1e-3);
        theta = result.first;
        num_iterations = result.second;

        end = std::chrono::high_resolution_clock::now();

        std::chrono::duration<double> em_time = end - start;

        std::cout << "Ran EM algorithm in " << em_time.count() << " seconds" << std::endl;
        std::cout << "Number of iterations: " << num_iterations << std::endl;

        std::cout << "Theta: " << std::endl;
        std::cout << theta << std::endl;
    } else {
        using T = float;

        std::vector<T> log_likelihood_matrix;
        std::vector<T> loglik_counts;

        auto start = std::chrono::high_resolution_clock::now();

        load_log_likelihoods_from_tsv(filename, log_likelihood_matrix, loglik_counts);

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> load_time = end - start;
        std::cout << "Loaded log likelihoods in " << load_time.count() << " seconds" << std::endl;

        torch::Tensor theta;
        int num_iterations;

        start = std::chrono::high_resolution_clock::now();

        auto result = em_algorithm(log_likelihood_matrix, loglik_counts, dtype);
        theta = result.first;
        num_iterations = result.second;

        end = std::chrono::high_resolution_clock::now();

        std::chrono::duration<double> em_time = end - start;

        std::cout << "Ran EM algorithm in " << em_time.count() << " seconds" << std::endl;
        std::cout << "Number of iterations: " << num_iterations << std::endl;

        std::cout << "Theta: " << std::endl;
        std::cout << theta << std::endl;
    }

    return 0;
}
